openmls/src/group/mls_group/mod.rs

```diff
@@ -229,31 +304,491 @@ impl MlsGroup {
+    ) -> Result<(&mut MessageSecrets, &[Member]), MessageDecryptionError> {
+        if epoch < self.context().epoch() {
+            self.message_secrets_store
+                .secrets_and_leaves_for_epoch_mut(epoch)
+                .ok_or({
+                    MessageDecryptionError::SecretTreeError(SecretTreeError::TooDistantInThePast)
+                })
+        } else {
+            // No need for leaves here. The tree of the current epoch is
+            // available to the caller.
+            Ok((self.message_secrets_store.message_secrets_mut(), &[]))
+        }
+    }
     /// Returns the leaf node of the client in the tree owning this group.
     pub fn own_leaf_node(&self) -> Option<&LeafNode> {
-        self.group.own_leaf_node().ok()
+        self.public_group().leaf(self.own_leaf_index())
     }
 
     /// Returns the group ID.
     pub fn group_id(&self) -> &GroupId {
-        self.group.group_id()
+        self.public_group.group_id()
     }
 
     /// Returns the epoch.
     pub fn epoch(&self) -> GroupEpoch {
-        self.group.context().epoch()
+        self.public_group.group_context().epoch()
+    }
+
+    /// Store the given [`EncryptionKeyPair`]s in the `provider`'s key store
+    /// indexed by this group's [`GroupId`] and [`GroupEpoch`].
+    ///
+    /// Returns an error if access to the key store fails.
+    pub(super) fn store_epoch_keypairs<Storage: StorageProvider>(
+        &self,
+        store: &Storage,
+        keypair_references: &[EncryptionKeyPair],
+    ) -> Result<(), Storage::Error> {
+        store.write_encryption_epoch_key_pairs(
+            self.group_id(),
+            &self.context().epoch(),
+            self.own_leaf_index().u32(),
+            keypair_references,
+        )
+    }
+
+    /// Read the [`EncryptionKeyPair`]s of this group and its current
+    /// [`GroupEpoch`] from the `provider`'s storage.
+    ///
+    /// Returns an empty vector if access to the store fails or it can't find
+    /// any keys.
+    pub(super) fn read_epoch_keypairs<Storage: StorageProvider>(
+        &self,
+        store: &Storage,
+    ) -> Vec<EncryptionKeyPair> {
+        store
+            .encryption_epoch_key_pairs(
+                self.group_id(),
+                &self.context().epoch(),
+                self.own_leaf_index().u32(),
+            )
+            .unwrap_or_default()
+    }
+
+    /// Delete the [`EncryptionKeyPair`]s from the previous [`GroupEpoch`] from
+    /// the `provider`'s key store.
+    ///
+    /// Returns an error if access to the key store fails.
+    pub(super) fn delete_previous_epoch_keypairs<Storage: StorageProvider>(
+        &self,
+        store: &Storage,
+    ) -> Result<(), Storage::Error> {
+        store.delete_encryption_epoch_key_pairs(
+            self.group_id(),
+            &GroupEpoch::from(self.context().epoch().as_u64() - 1),
+            self.own_leaf_index().u32(),
+        )
+    }
+
+    pub(crate) fn create_commit<Provider: OpenMlsProvider>(
```

Comment:
This should not be in `mod.rs`, it's distinct enough to have its own file.

Reply:
Fair enough. I've merged this function and the content of `create_commit_params.rs` into `create_commit.rs`.